### ISSUE TITLE
@evalpoly is type-unstable in certain situations

### ISSUE THREAD

**Issue by DumpsterDoofus** [JuliaLang/julia#10339]
This was originally posted [here](https://groups.google.com/forum/#!topic/julia-users/4GSW7C8vl1c), and Viral suggested filing as an issue. Here is a minimal example:

```
function test(a0::Float64, a1::Float64, c::Complex{Float64})
	for j in 1:1
		f = @evalpoly(c,a0, a1)
		if abs(f) < 1
			return f
		end
	end
	return 0.0im
end
typeof([test(1.0,2.0,x+y*im) for x in linspace(-1,1,2), y in linspace(-1,1,2)])
```
which returns `Array{Any,2}`. This causes significant gc and performance problems when constructing arrays. 

Strangely, when the outer `for` loop is removed, it becomes type-stable. I'm not sure what is going on here, but any clues would be appreciated.

**Comment by stevengj** [comment]
@JeffBezanson, is this some limitation of 98189bc6444898a341264bc6c02e9564b016e6db?

**Comment by ivarne** [comment]
I assume this fix is relevant for 0.3 as well.

**Comment by simonster** [comment]
Cherry picked in 6205d5ce2f03b81516b107f10fe3d150040dd662
